### PR TITLE
persistent dsu ys

### PR DESCRIPTION
**Pull request by envyaims**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which include the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making them.

### CODE REVIEW

File: solutions/ys-PersistentUnionFind.mdx [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+	vector<int> p, sz;
+	// stores info from the past
+	vector<pair<int,int>> past_parent, past_size;
+	DSU(int n) {
+		p.resize(n);
+		sz.resize(n,1);
+		iota(p.begin(),p.end(),0);
+	}
+	int get(int x) {
+		return p[x] == x ? x : get(p[x]);
```
Comment (Aadit-Ambadkar) [inline solutions/ys-PersistentUnionFind.mdx:33]:
```suggestion
		return ( p[x] == x ) ? x : get(p[x]);
```
clarity

File: solutions/ys-PersistentUnionFind.mdx [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+struct DSU{
+	vector<int> p, sz;
+	// stores info from the past
+	vector<pair<int,int>> past_parent, past_size;
+	DSU(int n) {
+		p.resize(n);
+		sz.resize(n,1);
+		iota(p.begin(),p.end(),0);
+	}
+	int get(int x) {
```
Comment (Aadit-Ambadkar) [inline solutions/ys-PersistentUnionFind.mdx:32]:
```suggestion
	int find(int x) {
```
traditionally named find, keep it as find.

File: solutions/ys-PersistentUnionFind.mdx (hunk 1 above)
Comment (Aadit-Ambadkar) [inline solutions/ys-PersistentUnionFind.mdx:33]:
```suggestion
		return p[x] == x ? x : find(p[x]);
```

File: solutions/ys-PersistentUnionFind.mdx [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+		return p[x] == x ? x : get(p[x]);
+	}
+	void unite(int a, int b) {
+		a = get(a); b = get(b);
```
Comment (Aadit-Ambadkar) [inline solutions/ys-PersistentUnionFind.mdx:36]:
```suggestion
		a = find(a); b = find(b);
```

File: solutions/ys-PersistentUnionFind.mdx [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+		past_parent.push_back({b, p[b]});
+		past_size.push_back({a, p[a]});
+		
+		if (a != b) {
+			p[b] = a;
+			sz[a] += sz[b];
+		}
+	}
+	bool sameset(int a, int b) {
+		return get(a) == get(b);
```
Comment (Aadit-Ambadkar) [inline solutions/ys-PersistentUnionFind.mdx:50]:
```suggestion
		return find(a) == find(b);
```

File: solutions/ys-PersistentUnionFind.mdx [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	vector<pair<int,int>> past_parent, past_size;
+	DSU(int n) {
+		p.resize(n);
+		sz.resize(n,1);
+		iota(p.begin(),p.end(),0);
+	}
+	int get(int x) {
+		return p[x] == x ? x : get(p[x]);
+	}
+	void unite(int a, int b) {
```
Comment (Aadit-Ambadkar) [inline solutions/ys-PersistentUnionFind.mdx:35]: **OPTIONAL**
I'd prefer if this was named merge, or union but unite works just fine

File: solutions/ys-PersistentUnionFind.mdx [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+		
+		if (a != b) {
+			p[b] = a;
+			sz[a] += sz[b];
+		}
+	}
+	bool sameset(int a, int b) {
+		return get(a) == get(b);
+	}
+	void rollback() {
```
Comment (Aadit-Ambadkar) [inline solutions/ys-PersistentUnionFind.mdx:52]:
```suggestion
	
	// Reverts to previous DSU state.
	void rollback() {
```

File: solutions/ys-PersistentUnionFind.mdx [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+	
+	vector<query> queries(q + 1);
+	vector<vector<int>> queries_using_k(q + 1);
+	
+	for (int i = 1; i <= q; i++) {
+		query tmp;
+		cin >> tmp.t >> tmp.k >> tmp.u >> tmp.v;
+		tmp.k++;
+		queries_using_k[tmp.k].push_back(i);
+		queries[i] = tmp;
```
Comment (Aadit-Ambadkar) [inline solutions/ys-PersistentUnionFind.mdx:77]: comments maybe?

File: solutions/ys-PersistentUnionFind.mdx [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+		// roll back merges
+		if (queries[idx].t == 0) {
+			dsu.rollback();
+		}
+	};
+	
+	dfs(dfs, 0);
+	
+	for (int i = 1; i <= q; i++) {
+		if(queries[i].t){
```
Comment (Aadit-Ambadkar) [inline solutions/ys-PersistentUnionFind.mdx:107]:
```suggestion
		if (queries[i].t){
```

**Review by Aadit-Ambadkar — CHANGES_REQUESTED** [review]
Very nice. Mostly just style changes and naming changes!

File: solutions/ys-PersistentUnionFind.mdx (hunk 2 above)
Comment (envyaims) [reply solutions/ys-PersistentUnionFind.mdx:32]: are you sure? cf edu uses get

File: solutions/ys-PersistentUnionFind.mdx (hunk 1 above)
Comment (envyaims) [reply solutions/ys-PersistentUnionFind.mdx:33]: wot

File: solutions/ys-PersistentUnionFind.mdx (hunk 7 above)
Comment (envyaims) [reply solutions/ys-PersistentUnionFind.mdx:79]: uh i think its pretty self explanatory?

File: solutions/ys-PersistentUnionFind.mdx [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+struct DSU{
```
Comment (SansPapyrus683) [inline solutions/ys-PersistentUnionFind.mdx:23]: spaces
```suggestion
struct DSU {
```

File: solutions/ys-PersistentUnionFind.mdx [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+	// stores info from the past
+	vector<pair<int,int>> past_parent, past_size;
+	DSU(int n) {
+		p.resize(n);
+		sz.resize(n,1);
+		iota(p.begin(),p.end(),0);
+	}
+	int get(int x) {
+		return ( p[x] == x ) ? x : get(p[x]);
+	}
```
Comment (SansPapyrus683) [inline solutions/ys-PersistentUnionFind.mdx:34]: add a newline between each function

File: solutions/ys-PersistentUnionFind.mdx [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+		sz[past_size.back().first] = past_size.back().second;
+		past_parent.pop_back();
+		past_size.pop_back();
+	}
+	
+};
+
+struct query{
+	int t, k, u, v;
+};
```
Comment (SansPapyrus683) [inline solutions/ys-PersistentUnionFind.mdx:65]: 1. rename to `Query`
2. spaces, yeah

File: solutions/ys-PersistentUnionFind.mdx [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+
+int main() {
+	cin.tie(0) -> sync_with_stdio(false);
+	int n, q; cin >> n >> q;
```
Comment (SansPapyrus683) [inline solutions/ys-PersistentUnionFind.mdx:69]: separate lines

File: solutions/ys-PersistentUnionFind.mdx (hunk 7 above)
Comment (SansPapyrus683) [reply solutions/ys-PersistentUnionFind.mdx:79]: "contruct our graph or whatever"

File: solutions/ys-PersistentUnionFind.mdx [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+	}
+	
+	vector<bool> ans(q + 1);
+	DSU dsu(n);
+	
+	auto dfs = [&](auto self, int idx) -> void {
+		if (queries[idx].t == 0) {
+			dsu.unite(queries[idx].u, queries[idx].v);
+		}
+		else{
```
Comment (SansPapyrus683) [inline solutions/ys-PersistentUnionFind.mdx:89]:
```suggestino
		} else {
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/ys-PersistentUnionFind.mdx (hunk 13 above)
Comment (envyaims) [reply solutions/ys-PersistentUnionFind.mdx:89]: i dont wanna be rude but like who does that

ill still add it though

File: solutions/ys-PersistentUnionFind.mdx (hunk 2 above)
Comment (SansPapyrus683) [reply solutions/ys-PersistentUnionFind.mdx:33]: yeah you can keep as is

File: solutions/ys-PersistentUnionFind.mdx [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+	// stores info from the past
+	vector<pair<int,int>> past_parent, past_size;
+	DSU(int n) {
+		p.resize(n);
+		sz.resize(n,1);
+		iota(p.begin(),p.end(),0);
+	}
+	
+	int get(int x) {
+		return ( p[x] == x ) ? x : get(p[x]);
```
Comment (SansPapyrus683) [inline solutions/ys-PersistentUnionFind.mdx:34]:
```suggestion
		return (p[x] == x) ? x : get(p[x]);
```

File: solutions/ys-PersistentUnionFind.mdx [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+		tmp.k++;
+		// construct query graph - add an edge between k and itself
+		queries_using_k[tmp.k].push_back(i); 
+		queries[i] = tmp;
+	}
+	
+	vector<bool> ans(q + 1);
+	DSU dsu(n);
+	
+	auto dfs = [&](auto self, int idx) -> void {
```
Comment (SansPapyrus683) [inline solutions/ys-PersistentUnionFind.mdx:89]: what this do

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/ys-PersistentUnionFind.mdx (hunk 15 above)
Comment (envyaims) [reply solutions/ys-PersistentUnionFind.mdx:89]: dfs down the query graph

Comment (SansPapyrus683) [reply solutions/ys-PersistentUnionFind.mdx:89]: add a comment here just to remind them

Comment (envyaims) [reply solutions/ys-PersistentUnionFind.mdx:89]: but... the function already says dfs?

Comment (SansPapyrus683) [reply solutions/ys-PersistentUnionFind.mdx:89]: no crap sherlock
i was asking for a comment like `// dfs to find the [whatever] or do the [whatever]`

Comment (envyaims) [reply solutions/ys-PersistentUnionFind.mdx:90]: theres really not much to say but ok

**Review by Aadit-Ambadkar — APPROVED** [review]
🫥👀

**Pull request merged**